man: Improve segment library documentation, provide examples

wenzeslaus opened the pull request:
This PR adds two examples into `doc/examples` which can be compiled as GRASS modules and executed. It adds 3 new sections to the library dox file: about general use with raster, about use with list of rasters (includes portions of one of the examples, so the code is only at one place), and a complete example (here including whole files). Finally, it modifies documentation for two functions.

Besides asking for general review of best practices in using this library, I'm asking these two questions:

* Is the remark about value initialization correct? (I don't really see a point in using calloc instead of malloc there considering that the non-memory case does initialize with zeros.)
* What about the precondition for `Segment_get_row()`? Does `Segment_get()` need the same one? `Segment_flush()` description suggests that only writing to disk and `Segment_get_row()` need that and `Segment_get_row()` does not call any page function if that has any influence.

wenzeslaus committed:
doc: examples for Segmement Library in dox and as code

wenzeslaus committed:
doc: advanced example for segmentlib in dox from example code

wenzeslaus committed:
doc: improved desc for open and get_row in segmentlib

metzm commented on doc/raster/r.example.segment/main.c:
```diff
@@ -0,0 +1,163 @@
+
+    /* determine the input map type (CELL/FCELL/DCELL) */
+    map_type = Rast_map_type(input_name, "");
+    size_t cell_size = Rast_cell_size(map_type);
```
move "size_t cell_size;" up, just after "RASTER_MAP_TYPE map_type;"

metzm commented on doc/raster/r.example.segment/main.c:
```diff
@@ -0,0 +1,163 @@
+    ncols = Rast_window_cols();
+
+    /* size of a segment */
+    int srows = 64;
```
Move "int srows, scols, nseg;" up

metzm commented on doc/raster/r.example.segment/main.c:
```diff
@@ -0,0 +1,163 @@
+    int num_seg = 4;
+
+    /* segment structure */
+    SEGMENT raster_seg;
```
move "SEGMENT raster_seg;" up

metzm commented on doc/raster/r.example.segmulti/main.c:
```diff
@@ -0,0 +1,189 @@
+
+    /* pass the pointer, set the value */
+    Segment_put(raster_seg, values, row, col);
+}
```
+ G_free(values):

metzm replied:
+ G_free(values);

metzm commented on doc/raster/r.example.segmulti/main.c:
```diff
@@ -0,0 +1,189 @@
+    ncols = Rast_window_cols();
+
+    /* size of a segment */
+    int srows = 64;
```
as for r.example.segment, move int srows, sclos, nsegs; up

metzm commented on doc/raster/r.example.segmulti/main.c:
```diff
@@ -0,0 +1,189 @@
+    int nsegs = 4;
+
+    map_type = DCELL_TYPE;
+    size_t cell_size = Rast_cell_size(map_type);
```
Move "size_t cell_size, segment_cell_size;" up

metzm commented on doc/raster/r.example.segmulti/main.c:
```diff
@@ -0,0 +1,189 @@
+    output_fd = Rast_open_new(output_name, map_type);
+
+    /* segment structure */
+    SEGMENT raster_seg;
```
move up

metzm commented on doc/raster/r.example.segmulti/main.c:
```diff
@@ -0,0 +1,189 @@
+        G_fatal_error(_("Unable to create temporary segment file"));
+
+    /* array to store file descriptors */
+    int *input_fds = G_malloc(ninputs * sizeof(int));
```
move "int *input_fds;" up

metzm commented on doc/raster/r.example.segmulti/main.c:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    /* allocate input buffer */
+    DCELL *row_buffer = Rast_allocate_d_buf();
```
move "DCELL *row_buffer, *seg_buffer;" up

metzm commented on doc/raster/r.example.segmulti/r.example.segmulti.html:
```diff
@@ -0,0 +1,44 @@
+
+<em>
+<a href="r.example.html">r.example</a>
+<a href="r.example.html">r.example.segment</a>
```
r.example.segment.html

metzm commented on lib/segment/get_row.c:
```diff
@@ -21,17 +21,22 @@
- * <b>Buf</b> will be filled with <em>ncols*len</em> bytes of data
- * corresponding to the <b>row</b> in the data matrix.
+ * \pre *buf* points to an allocated array of length <em>ncols*len</em>
+ * where *len* is size of one stored value (cell).
```
e.g. cell or a custom data structure

metzm commented on lib/segment/open.c:
```diff
@@ -21,10 +21,28 @@
+ * Segment_init(), all-in-memory mode initialization, and temporary
+ * file handling.
+ *
+ * The values are not guaranteed to be initialized to zero or NULL.
```
not values, but bytes, they are indeed initialized to zero because of lseek and calloc

metzm commented on lib/segment/segmentlib.dox:
```diff
@@ -64,21 +69,253 @@ or undocumented, start with the prefix \c Segment_. To avoid name
+the raster you need to access:
+
+\code
+for (int row = 0; row < Rast_window_rows(); row++) {
```
for (int row = 0) is C++ style

metzm commented on lib/segment/segmentlib.dox:
```diff
@@ -64,21 +69,253 @@ or undocumented, start with the prefix \c Segment_. To avoid name
+
+\skip G_free
+\until Rast_close(output_fd);
+
```
+ Segment_close() ?

metzm commented:
> This PR adds two examples into `doc/examples` which can be compiled as GRASS modules and executed. It adds 3 new sections to the library dox file: about general use with raster, about use with list of rasters (includes portions of one of the examples, so the code is only at one place), and a complete example (here including whole files). Finally, it modifies documentation for two functions.

Please use standard C style in sample code.
Maybe documentation does not need to be that long? avoid tl;dr
> 
> Besides asking for general review of best practices in using this library, I'm asking these two questions:
> 
>     * Is the remark about value initialization correct? (I don't really see a point in using calloc instead of malloc there considering that the non-memory case does initialize with zeros.)
See "man calloc". malloc does not initialize the memory while calloc does.
> 
>     * What about the precondition for `Segment_get_row()`? Does `Segment_get()` need the same one?
No.

> `Segment_flush()` description suggests that only writing to disk and `Segment_get_row()` need that and `Segment_get_row()` does not call any page function if that has any influence.

Yes, that's the reason: Segment_get_row() bypasses the internal cache and reads directly from the temporary file.

metzm commented:
Considering examples on how to use the segment library, it is IMHO sufficient to mention current working modules like r.cost, r.walk, r.watershed, r.stream.extract.

mlennert commented:
Jumping into this discussion a bit late, so just two short comments:

- I think we definitely need enhanced and detailed documentation of the C-code. @metzm, today you are probably the only one to really know the nitty-gritty stuff in this library, so for future sustainability (and possibly attracting the rare C-developer out there), good doc seems a must to me.

- At this stage, I haven't had the chance to delve into the doc enough to understand whether new (simple) examples are needed. They have the advantage of focusing on the features documented, thus avoiding distraction through other unrelated parts of code. They have the disadvantage of having to be updated when the library changes. Actual production modules obviously provide more motivation to keep them up to date.

I don't think this has been done as such in the code doc, yet, but maybe we could just add specific references to the existing production modules directly into the doc ?

wenzeslaus commented:
Note to myself: It seems that it worth explicitly noting that functions such as `Segment_get()` don't do bounds check for row and col (and thus crash will occur rather than an error message).

neteler commented:
@wenzeslaus would you mind to rebase this PR?

echoix committed:
Merge branch 'main' into segment-lib-doc

echoix commented:
@wenzeslaus I manually resolved the conflicts

echoix committed:
Apply suggestions from clang-format

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

echoix committed:
Merge branch 'main' into segment-lib-doc

echoix committed:
Merge branch 'main' into segment-lib-doc

echoix committed:
Merge branch 'main' into segment-lib-doc

echoix commented on doc/raster/r.example.segment/README.md:
````diff
@@ -0,0 +1,20 @@
+To compile the example simply use `make` (in this directory):
+
+```
+make
+```
````
````suggestion
```shell
make
```
````

echoix commented on doc/raster/r.example.segment/README.md:
````diff
@@ -0,0 +1,20 @@
+To run (the asterisks will match your operating system and version
+specific directory and file):
+
+```
+../../../bin.*/grass* --tmp-location XY --exec bash <<EOF
+    g.region res=0.1
+    r.mapcalc -s expression='raster_map_1 = rand(0., 15)'
+    r.example.segment input=raster_map_1 output=raster_map_2
+    r.univar raster_map_1
+    r.univar raster_map_2
+EOF
+```
````
````suggestion
```shell
../../../bin.*/grass* --tmp-location XY --exec bash <<EOF
    g.region res=0.1
    r.mapcalc -s expression='raster_map_1 = rand(0., 15)'
    r.example.segment input=raster_map_1 output=raster_map_2
    r.univar raster_map_1
    r.univar raster_map_2
EOF
```
````

echoix commented on doc/raster/r.example.segmulti/README.md:
````diff
@@ -0,0 +1,34 @@
+To run (the asterisks will match your operating system and version
+specific directory and file):
+
+```
+../../../bin.*/grass* --tmp-location XY --exec bash <<EOF
+g.region res=0.1
+r.mapcalc -s expression='raster_1 = rand(0., 15)'
+r.mapcalc -s expression='raster_2 = rand(0., 15)'
+r.mapcalc -s expression='raster_3 = rand(0., 15)'
+r.example.segmulti input=raster_1,raster_2,raster_3 output=raster_out
+r.univar raster_1
+r.univar raster_2
+r.univar raster_3
+r.univar raster_out
+r.info -g raster_out
+r.describe raster_out
+g.gui -f
+EOF
+
+```
````
````suggestion
```shell
../../../bin.*/grass* --tmp-location XY --exec bash <<EOF
g.region res=0.1
r.mapcalc -s expression='raster_1 = rand(0., 15)'
r.mapcalc -s expression='raster_2 = rand(0., 15)'
r.mapcalc -s expression='raster_3 = rand(0., 15)'
r.example.segmulti input=raster_1,raster_2,raster_3 output=raster_out
r.univar raster_1
r.univar raster_2
r.univar raster_3
r.univar raster_out
r.info -g raster_out
r.describe raster_out
g.gui -f
EOF

```
````

echoix commented on doc/raster/r.example.segmulti/README.md:
````diff
@@ -0,0 +1,34 @@
+
+To precisely time the execution, you can use *perf*:
+
+```
+perf stat -r 100 r.example.segmulti ... --overwrite
+```
````
````suggestion
```shell
perf stat -r 100 r.example.segmulti ... --overwrite
```
````

echoix committed:
Add language in markdown fenced code blocks

echoix committed:
Merge branch 'main' into segment-lib-doc

echoix commented on doc/raster/r.example.segmulti/README.md:
````diff
@@ -0,0 +1,34 @@
+To compile the example simply use `make` (in this directory):
+
+```
+make
+```
````
````suggestion
```shell
make
```
````

echoix committed:
Update README.md

echoix reviewed (commented):
@wenzeslaus is this still relevant now? It is currently in a state where it up to date enough to pass CI, but there are still some very old (6 years) review comments.

Also, could you take a new look, to tell us if it still shows the patterns we want to show as an example?

It's kind of funny to still see the issue/pr number #20, that is still open. Probably the oldest we have open. _Edit: it **is** the oldest open PR or issue we have_

echoix reviewed (changes requested):
It is our oldest open PR, number 20. We have to port the docs changes in HTML to markdown. Also, since the last update of the PR, the folder structure changed (in early January 2025, in #4815)

echoix committed:
Merge branch 'main' into segment-lib-doc

echoix committed:
Merge branch 'main' into segment-lib-doc

wenzeslaus committed:
Merge remote-tracking branch 'upstream/main' into segment-lib-doc

wenzeslaus committed:
doc: Update segment library examples to current layout and Markdown

Move r.example.segment and r.example.segmulti from doc/raster to
doc/examples/raster to follow the directory structure introduced
in #4815, and adjust the Makefiles and READMEs for the new depth.

Add Markdown documentation files as the canonical documentation and
keep the HTML files in sync with them. Fix the cross-links in the
SEE ALSO sections (review comment), use the current programming
manual URL, link the Segment Library page directly, use the map
names from the example in the r.example.segment EXAMPLE section,
and drop the obsolete CVS date keyword. Use --tmp-project in the
READMEs to match the current command line interface.

In the example code, process the raster as DCELL regardless of the
input map type. Previously, the segment cell size followed the
input map type while process() assumed DCELL, so for CELL and FCELL
inputs the modification was silently lost (verified by running the
tool before and after the change). Free the value buffer in
r.example.segmulti (review comment), allocate it with sizeof(DCELL)
instead of sizeof(DCELL *), and mark a message for translation.

In the library manual page, point the example includes to the new
paths, mention Segment_close() in the cleanup step (review comment),
show the buffer deallocation in the processing extract, rename the
complete example section to match its content (one input raster),
and fix typos. Use an absolute path for the doxygen EXAMPLE_PATH so
the examples are found both from the top directory and from
per-directory runs such as make -C lib/segment htmldox.

Update prepared with AI assistance (Claude).

wenzeslaus committed:
CI: Compile segment library example modules

Compile r.example.segment and r.example.segmulti in the step which
compiles the other example modules, so the examples stay buildable
without being part of the regular build.

Update prepared with AI assistance (Claude).

wenzeslaus committed:
lib/segment: Clarify Segment_open and Segment_get_row docs

Address review comments on the changed function documentation:

In Segment_open(), state that the underlying bytes read as zeros in
both the all-in-memory mode (G_calloc) and the temporary file mode
(file created by seeking), instead of the incorrect claim that no
initialization is guaranteed, while still noting that values are
not initialized to raster NULL values. Describe when the
all-in-memory mode is used instead of the vague automatic
adjustment wording, and attribute the out of memory fatal error to
the underlying G_calloc(). Move the comment about the mode marker
to the cache allocation because the library detects the
all-in-memory mode by the non-NULL cache, not by scb.

In Segment_get_row(), extend the description of the stored value
size with examples (review comment) and smooth the wording.

Update prepared with AI assistance (Claude).
